### PR TITLE
New Resource: `azurerm_redhat_openshift_cluster`

### PR DESCRIPTION
**Pull request by mbfrahry**
Supersedes #15262

### CODE REVIEW

**Comment by david-yu** [comment]
Thank you, looking forward to this PR!

**Comment by david-yu** [comment]
Looks like this is where the latest sdk lives: https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/resourcemanager/redhatopenshift/armredhatopenshift

**Comment by mbfrahry** [comment]
@yukirii, do you happen to know how to find the specific error for RedHatOpenShift failures? I keep getting `Internal server error.` But there isn't any information outside of that in the portal.

**Comment by yukirii** [comment]
sorry I don't have information how to find the error.. 
`Internal server error` is a common issue in ARO when a cluster deployment is failed or if there are in-sufficient permissions for the ARO Resource Provider on the deployed resources or policy restrictions on the Azure Resources. Resource Provider returns only `Internal server error` message, it is need to open support ticket to figure out root cause.

**Comment by yukirii** [comment]
@mbfrahry Can you share .tf files (or commands) you used for testing? I'll take a look if there are any points that could cause an error.

**Comment by mbfrahry** [comment]
Hey @yukirii, I'm basing this off of the [basic test](https://github.com/hashicorp/terraform-provider-azurerm/pull/20266/files#diff-9ba9a32872687bcd4f29121a0c8ab770fd3d47239a96c2c5b4d41419d57e1d54R112). I also can't delete what I've tested. I've got about 7 that have failed creation but every time I delete one, it comes back a few minutes later. Even deleting the resource group isn't working. I'll open a support ticket to try and get those deleted and to find out some more information. 

Do you know if there is any plan to surface more information about `Internal Server Errors`?  I'm not sure we can support this resource in its current form if we can't get better errors so people know exactly why their configuration is failing.

**Comment by katbyte** [comment]
@yukirii - is there any you could update the service to correctly bubble up errors? without knowing what is actually causing an error it makes both development and continual maintenance of this resource difficult, if not imposssible. Not to mention users not know how to resolve issues they have during deployment.

**Comment by mzarglis** [comment]
I ran into this exact same thing when writing module for ARO using AzAPI and got the same response from Microsoft about the only way to get the actual error is to open up a support ticket 😬 In my case it was the cluster spn oauth callback uri being incorrect. https://github.com/Azure-Samples/aro-azapi-terraform could be useful to cross reference your implementation...

**Comment by mbfrahry** [comment]
Thanks for that additional info @mzarglis. We'll have to look into it for sure but I'm hesitant to even release this if it's something that people will continue to run into without any pointers from the api to tell them what exactly is going on. I've opened an issue on their [api-specs](https://github.com/Azure/azure-rest-api-specs/issues/22514) and hopefully we can get some traction there. 

For now, I'm going to have to say this is blocked

**Comment by mzarglis** [comment]
> Thanks for that additional info @mzarglis. We'll have to look into it for sure but I'm hesitant to even release this if it's something that people will continue to run into without any pointers from the api to tell them what exactly is going on. I've opened an issue on their [api-specs](https://github.com/Azure/azure-rest-api-specs/issues/22514) and hopefully we can get some traction there.
> 
> For now, I'm going to have to say this is blocked

Makes sense looking forward to the official azurerm implementation when these issues are sorted. Your work on this is greatly appreciated!

**Comment by yukirii** [comment]
@mbfrahry Thank you for replying. Is the following command correct to run basic test for the resource?
```
make acctests SERVICE='redhatopenshift'
```
https://github.com/hashicorp/terraform-provider-azurerm/blob/main/DEVELOPER.md#developing-the-provider

**Comment by mbfrahry** [comment]
Hey @yukirii, apologies for the delay. The command I am running is `make testacc TEST='./internal/services/redhatopenshift/' TESTARGS='-run=TestAccOpenShiftCluster_basic1'`

**Comment by yukirii** [comment]
Hi @mbfrahry, thank you for providing command. I'm trying to run test using `TF_LOG=DEBUG make testacc TEST='./internal/services/redhatopenshift/' TESTARGS='-run=TestAccOpenShiftCluster_basic1'`.

The first error I encountered is invalid domain name. `The provided domain '774mg.com' is invalid.`

```
2023-02-20T20:47:52.430+0900 [DEBUG] sdk.helper_resource: Started sdkv2 provider instance server: test_name=TestAccOpenShiftCluster_basic1 test_step_number=1 test_terraform_path=/home/yukirii/bin/terraform test_working_directory=/tmp/plugintest3053176881 tf_provider_addr=registry.terraform.io/hashicorp/azurerm-alt
2023-02-20T20:48:30.986+0900 [WARN]  sdk.helper_resource: Error running Terraform CLI command: test_name=TestAccOpenShiftCluster_basic1 test_step_number=1 test_terraform_path=/home/yukirii/bin/terraform
  error=
  | exit status 1
  | 2023/02/20 20:47:52 [DEBUG] Using modified User-Agent: Terraform/0.12.20 HashiCorp-terraform-exec/0.17.3
  |
  | Error: creating /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/acctestRG-aro-230220204731285525/providers/Microsoft.RedHatOpenShift/openShiftClusters/acctestaro230220204731285525: performing CreateOrUpdate: openshiftclusters.OpenShiftClustersClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidParameter" Message="The provided domain '774mg.com' is invalid." Target="properties.clusterProfile.domain"
  |
  |   on terraform_plugin_test.tf line 89, in resource "azurerm_redhat_openshift_cluster" "test":
  |   89: resource "azurerm_redhat_openshift_cluster" "test" {
```

[RandString()](https://github.com/hashicorp/terraform-plugin-testing/blob/82ad008d3de1af6976226ae539d5ac8c6b56dff3/helper/acctest/random.go#L48) returns string which contains digits. But ARO Resource Provider doesn't accept domain which start with a digit. We need to use domain which starts with non-digit character when testing.
https://github.com/Azure/ARO-RP/blob/84b6e0777843186ed3287df3743c0099fe2b3367/pkg/api/v20220904/openshiftcluster_validatestatic.go#L136-L141

I'll also look into regarding InternalServerError.

File: website/docs/r/redhatopenshift_cluster.html.markdown [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+---
+
+A `service_principal` block supports the following:
+
+* `client_id` - (Required) The Client ID for the Service Principal. Changing this forces a new resource to be created.
+
+* `client_secret` - (Required) The Client Secret for the Service Principal. Changing this forces a new resource to be created.
+
+~> **Note:** Currently a service principal cannot be associated with more than one ARO clusters on the Azure subscription.
```
Comment (guptaaayush8) [inline website/docs/r/redhatopenshift_cluster.html.markdown:140]: This service principal is also given Contributor access to ARO, by default, when creating via azure portal.

**Review by yukirii — COMMENTED** [review]
Adding some changes to test data, ARO cluster creation succeeded without `InternalServerError` by `make testacc TEST='./internal/services/redhatopenshift/' TESTARGS='-run=TestAccOpenShiftCluster_basic1'`.

But one check failed. We need fix the check and other templates of cluster.
`2023-03-05T16:24:12.370+0900 [ERROR] sdk.helper_resource: Unexpected error: test_terraform_path=/home/user/bin/terraform test_working_directory=/tmp/plugintest2521753468 error="Check failed: Check 5/7 error: azurerm_redhat_openshift_cluster.test: Attribute 'worker_profile.0.node_count' expected \"3\", got \"1\"" test_name=TestAccOpenShiftCluster_basic1 test_step_number=1    testcase.go:110: Step 1/1 error: Check failed: Check 5/7 error: azurerm_redhat_openshift_cluster.test: Attribute 'worker_profile.0.node_count' expected "3", got "1"`

Also, I confirmed `terraform apply` and `terraform destroy` succeeded with the following tf file.
https://gist.github.com/yukirii/30fc44a7686a1632cef10431bef950fd

**Comment by guptaaayush8** [comment]
> Adding some changes to test data, ARO cluster creation succeeded without `InternalServerError` by `make testacc TEST='./internal/services/redhatopenshift/' TESTARGS='-run=TestAccOpenShiftCluster_basic1'`.
> 
> But one check failed. We need fix the check and other templates of cluster. `2023-03-05T16:24:12.370+0900 [ERROR] sdk.helper_resource: Unexpected error: test_terraform_path=/home/user/bin/terraform test_working_directory=/tmp/plugintest2521753468 error="Check failed: Check 5/7 error: azurerm_redhat_openshift_cluster.test: Attribute 'worker_profile.0.node_count' expected \"3\", got \"1\"" test_name=TestAccOpenShiftCluster_basic1 test_step_number=1 testcase.go:110: Step 1/1 error: Check failed: Check 5/7 error: azurerm_redhat_openshift_cluster.test: Attribute 'worker_profile.0.node_count' expected "3", got "1"`
> 
> Also, I confirmed `terraform apply` and `terraform destroy` succeeded with the following tf file. https://gist.github.com/yukirii/30fc44a7686a1632cef10431bef950fd

ARO requires a minimum of 3 worker nodes and only 3 master nodes. it is a requirement from Azure portal as well.

**Comment by guptaaayush8** [comment]
@mbfrahry @david-yu @yukirii - Hey! is there any update on the module?

**Comment by qazimobeen** [comment]
Hi Team, are we progressing on it further? What other alternative do exist for this? 

Shall we fallback on Ansible/Bicep way of working?

**Comment by tombuildsstuff** [comment]
👋  hey folks

Since this one's been sitting for a while and continues to be blocked by the API we're going to close this PR for the moment - it's worth noting that this is also using the older `Azure/azure-sdk-for-go` rather than `hashicorp/go-azure-sdk` and as such would need to be updated to use that SDK.

Whilst we're going to close this PR if someone else is interested in picking this one up we'd be open to that, however given the current issues with the API (insofar as it returns more detail than `Internal Server Error` i.e. a description of the problem) I believe this remains blocked for the moment?

Thanks!

**Pull request closed without merging**